lib/contents.php

```diff
@@ -1,244 +1,133 @@
-	private $responseCode;
-	private $responseHeaders;
-	private $responseBody;
-
-	protected function getMessageHeading() {
-		return 'Unexpected response from upstream';
-	}
-
-	public function __construct($responseBody, $responseHeaders, $responseCode = 500, Throwable $previous = null) {
-		$this->responseCode = $responseCode;
-		$this->responseHeaders = $responseHeaders;
-		$this->responseBody = $responseBody;
-
-		parent::__construct('', $responseCode, $previous);
+final class HttpException extends \Exception {}
+
+function getContents(
+	string $url,
+	array $httpHeaders = [],
+	array $curlOptions = [],
+	bool $returnHeader = false
+) {
+	$cacheFactory = new CacheFactory();
+	$cacheFactory->setWorkingDir(PATH_LIB_CACHES);
+	$cache = $cacheFactory->create(Configuration::getConfig('cache', 'type'));
+	$cache->setScope('server');
+	$cache->purgeCache(86400); // 24 hours (forced)
+	$cache->setKey([$url]);
+
+	$config = [
+		'headers' => $httpHeaders,
+		'curl_options' => $curlOptions,
+	];
+	if (defined('PROXY_URL') && !defined('NOPROXY')) {
+		$config['proxy'] = PROXY_URL;
 	}
-
-	public function getResponseCode() {
-		return $this->responseCode;
+	if(!Debug::isEnabled() && $cache->getTime()) {
+		$config['if_not_modified_since'] = $cache->getTime();
 	}
 
-	public function getResponseHeaders() {
-		return $this->responseHeaders;
+	$result = _http_request($url, $config);
+	$response = [
+		'header' => $result['headers'],
+		'content' => $result['body'],
+	];
+
+	switch($result['code']) {
+		case 200:
+		case 201:
+		case 202:
+			if(isset($result['headers']['cache-control'])) {
+				$cachecontrol = $result['headers']['cache-control'];
+				$lastValue = array_pop($cachecontrol);
+				$directives = explode(',', $lastValue);
+				$directives = array_map('trim', $directives);
+				if(in_array('no-cache', $directives) || in_array('no-store', $directives)) {
+					break;
+				}
+			}
+			$cache->saveData($result['body']);
+			break;
+		case 304: // Not Modified
+			$response['content'] = $cache->loadData();
+			break;
+		default:
+			throw new HttpException('', $result['code']);
 	}
-
-	public function getResponseBody() {
-		return $this->responseBody;
+	if ($returnHeader === true) {
+		return $response;
 	}
+	return $response['content'];
 }
 
 /**
- * Gets contents from the Internet.
- *
- * **Content caching** (disabled in debug mode)
- *
- * A copy of the received content is stored in a local cache folder `server/` at
- * {@see PATH_CACHE}. The `If-Modified-Since` header is added to the request, if
- * the provided URL has been cached before.
- *
- * When the server responds with `304 Not Modified`, the cached data is returned.
- * This will improve response times and reduce bandwidth for servers that support
- * the `If-Modified-Since` header.
- *
- * Cached files are forcefully removed after 24 hours.
- *
- * @link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since
- * If-Modified-Since
```

Comment:
Less is more. It was a wall of text that didn't have much value. Maybe add back the part about the debug/cache behavior. It was lying though, it still did http requests even if debug was enabled.

Reply:
Your PR completely removes the docstring for getContents. The comments on the params and return types are useful, especially with the external links.

>  It was lying though, it still did http requests even if debug was enabled.

I don't see where it implies that no requests will be made when debug mode is enabled.

Reply:
The link to MDN for If-Modified-Since was very much malplaced. It is an implementation detail. Why would we want that in the function comment?

Return type is actually array|string. The array part is used on a few places where the response headers are wanted.

Ah yes you are right. The intended behaviour is to do http requests even when debug is disabled. I was thinking it was a bug and that it should not do http request when we have cached data.

Reply:
> The link to MDN for If-Modified-Since was very much malplaced. It is an implementation detail. Why would we want that in the function comment?

Then wouldn't it be better to move it to the implementation location rather than completely deleting it?

> Return type is actually array|string. The array part is used on a few places where the response headers are wanted.

Isn't that what is documented?
```
 * @return string|array The contents.
```

Reply:
I think I had a plan to drop that and only return the response data. Or maybe a response object. As far as I can see, no bridges uses the response headers for anything. It's very rarely used. I'll add back the phpdoc. I think you are placing way too much value in comments. Honest disagreement there.

Reply:
This breaks the connectivity action, which uses the headers to search for redirects. Could you restore the original behavior with returnHeaders=true?